File: web/presentation.js
const faceOrder={left:0,right:1,front:2,back:3,bottom:4,top:5};

export function displayNetName(design,netId){
  return design?.nets?.find(net=>net.id===netId)?.label||netId||'Unassigned';
}

function stableFeatureOrder(a,b){
  return (faceOrder[a.face]??99)-(faceOrder[b.face]??99)||
    (a.u??0)-(b.u??0)||(a.v??0)-(b.v??0)||(a.depth??0)-(b.depth??0)||String(a.id).localeCompare(String(b.id));
}

export function displayExternalPortName(feature,design){
  if(feature.schematic_id)return feature.schematic_id;
  if(feature.id&&!String(feature.id).startsWith('PORT_'))return feature.id;
  const peers=(design?.features||[]).filter(item=>item.kind==='port'&&item.circuit===feature.circuit&&!item.suppressed).sort(stableFeatureOrder);
  const net=displayNetName(design,feature.circuit),index=peers.findIndex(item=>item.id===feature.id);
  return peers.length<=1?net:`${net}${Math.max(0,index)+1}`;
}

export function displayInterfaceName(design,featureId,interfaceId,{definitionOnly=false,index=null}={}){
  const feature=(design?.features||[]).find(item=>item.id===featureId);
  const net=feature?.interface_nets?.[interfaceId];
  if(feature&&net)return `${displayFeatureName(feature,design)}-${displayNetName(design,net)}`;
  const n=index??Math.max(0,Object.keys(feature?.interface_nets||{}).indexOf(interfaceId));
  return definitionOnly?`Interface ${n+1} · ${interfaceId}`:`${displayFeatureName(feature||{id:featureId,kind:'cavity'},design)} · Interface ${n+1}`;
}

export function displayRouteName(feature,design){
  const netId=feature.route_net||feature.frozen_net;
  if(feature.kind!=='drilling'||!netId)return null;
  const features=design?.features||[];
  const route=features.filter(item=>item.kind==='drilling'&&(item.route_net||item.frozen_net)===netId).sort(stableFeatureOrder);
  const referencedOwners=new Set(route.flatMap(item=>item.connects_to||[]).filter(value=>String(value).includes(':')).map(value=>String(value).split(':')[0]));
  const owners=features.filter(item=>item.kind==='cavity'&&!item.suppressed&&(Object.values(item.interface_nets||{}).includes(netId)||referencedOwners.has(item.id))).sort(stableFeatureOrder);
  const prefix=owners.length===1?`${displayFeatureName(owners[0],design)}-${displayNetName(design,netId)}`:displayNetName(design,netId);
  return `${prefix}${Math.max(0,route.findIndex(item=>item.id===feature.id))+1}`;
}

export function displayFeatureName(feature,design){
  if(!feature)return 'Unknown feature';
  const route=displayRouteName(feature,design);if(route)return route+(feature.plugged?' · PLUG':'');
  if(feature.kind==='port')return displayExternalPortName(feature,design);
  return design?.schematic_intent?.components?.find(component=>component.placement_id===feature.id)?.label||feature.id;
}

export function displayMemberName(design,member){
  const text=String(member||'');
  const split=text.indexOf(':');
  if(split>0){
    const owner=text.slice(0,split),suffix=text.slice(split+1);
    const feature=(design?.features||[]).find(item=>item.id===owner);
    if(feature?.kind==='cavity')return displayInterfaceName(design,owner,suffix);
  }
  const feature=(design?.features||[]).find(item=>item.id===text);
  return feature?displayFeatureName(feature,design):text;
}

const ruleNames={
  engineering_review:'Engineering review',schematic_conformance:'Schematic conformance',
  cartridge_compatibility:'Cartridge compatibility',connected_interface:'Hydraulic interface connection',
  circuit_connectivity:'Hydraulic net connectivity',minimum_feature_wall:'Minimum feature wall',
  external_wall:'External wall',solid_validity:'Solid validity',solid_count:'Solid count',
  step_round_trip:'STEP round trip',cavity_collision:'Cavity collision',pressure_strength:'Pressure strength'
};
export function displayRuleName(rule){return ruleNames[rule]||String(rule||'').replaceAll('_',' ').replace(/^./,c=>c.toUpperCase());}

export function displayReviewName(value){
  const text=String(value||'');
  if(text==='PORT_SPEC')return 'External port specification';
  if(text==='External ports')return 'External ports';
  return text.startsWith('REVIEW_')?'Engineering review item':text;
}

export function displayIdentity(design,value){
  let text=String(value??'');
  if(text.startsWith('F:'))text=text.slice(2).split(':step:')[0];
  const feature=(design?.features||[]).find(item=>item.id===text);
  if(feature)return displayFeatureName(feature,design);
  const split=text.indexOf(':');
  if(split>0){
    const owner=text.slice(0,split),suffix=text.slice(split+1);
    const placed=(design?.features||[]).find(item=>item.id===owner);
    if(placed?.kind==='cavity')return displayInterfaceName(design,owner,suffix);
  }
  return displayReviewName(text);
}

File: manifold/presentation.py
"""Human-facing names for stable engineering identities.

Machine artifacts keep the original IDs.  This module is only for UI, drawings
and human-readable reports.
"""
import re


_FACE_ORDER={'left':0,'right':1,'front':2,'back':3,'bottom':4,'top':5}
_RULE_NAMES={
    'engineering_review':'Engineering review','schematic_conformance':'Schematic conformance',
    'cartridge_compatibility':'Cartridge compatibility','connected_interface':'Hydraulic interface connection',
    'circuit_connectivity':'Hydraulic net connectivity','minimum_feature_wall':'Minimum feature wall',
    'external_wall':'External wall','solid_validity':'Solid validity','solid_count':'Solid count',
    'step_round_trip':'STEP round trip','cavity_collision':'Cavity collision','pressure_strength':'Pressure strength',
}


def _features(design):
    return list(design.features if hasattr(design,'features') else design.get('features',[]))


def _get(value,key,default=None):
    return getattr(value,key,value.get(key,default) if isinstance(value,dict) else default)


def net_name(design,net_id):
    nets=design.nets if hasattr(design,'nets') else design.get('nets',[])
    row=next((n for n in nets if _get(n,'id')==net_id),None)
    return _get(row,'label') or net_id or 'Unassigned'


def _order(feature):
    return (_FACE_ORDER.get(_get(feature,'face'),99),_get(feature,'u',0) or 0,_get(feature,'v',0) or 0,
            _get(feature,'depth',0) or 0,str(_get(feature,'id','')))


def feature_name(design,feature):
    if feature is None:return 'Unknown feature'
    route=_get(feature,'route_net') or _get(feature,'frozen_net')
    if _get(feature,'kind')=='drilling' and route:
        peers=sorted((f for f in _features(design) if _get(f,'kind')=='drilling' and (_get(f,'route_net') or _get(f,'frozen_net'))==route),key=_order)
        referenced={str(target).split(':',1)[0] for peer in peers for target in (_get(peer,'connects_to',[]) or []) if ':' in str(target)}
        owners=sorted((f for f in _features(design) if _get(f,'kind')=='cavity' and not _get(f,'suppressed',False)
                       and (route in (_get(f,'interface_nets',{}) or {}).values() or _get(f,'id') in referenced)),key=_order)
        prefix=f'{feature_name(design,owners[0])}-{net_name(design,route)}' if len(owners)==1 else net_name(design,route)
        label=f'{prefix}{max(0,next((i for i,f in enumerate(peers) if _get(f,"id")==_get(feature,"id")),0))+1}'
        return label+(' · PLUG' if _get(feature,'plugged',False) else '')
    if _get(feature,'kind')=='port':
        if _get(feature,'schematic_id'):return _get(feature,'schematic_id')
        if _get(feature,'id') and not str(_get(feature,'id')).startswith('PORT_'):return _get(feature,'id')
        peers=sorted((f for f in _features(design) if _get(f,'kind')=='port' and _get(f,'circuit')==_get(feature,'circuit') and not _get(f,'suppressed',False)),key=_order)
        label=net_name(design,_get(feature,'circuit'))
        return label if len(peers)<=1 else f'{label}{max(0,next((i for i,f in enumerate(peers) if _get(f,"id")==_get(feature,"id")),0))+1}'
    components=getattr(design,'components',[])
    component=next((c for c in components if _get(c,'feature_id')==_get(feature,'id')),None)
    return _get(component,'label') or _get(feature,'id')


def interface_name(design,feature_id,interface_id,*,definition_only=False,index=0):
    feature=next((f for f in _features(design) if _get(f,'id')==feature_id),None)
    net=(_get(feature,'interface_nets',{}) or {}).get(interface_id) if feature else None
    if feature and net:return f'{feature_name(design,feature)}-{net_name(design,net)}'
    return f'Interface {index+1} · {interface_id}' if definition_only else f'{feature_id} · Interface {index+1}'


def review_name(value):
    text=str(value or '')
    if text=='PORT_SPEC':return 'External port specification'
    if text.startswith('REVIEW_'):return 'Engineering review item'
    return text


def identity_name(design,value):
    text=str(value or '')
    if text.startswith('F:'):text=text[2:].split(':step:')[0]
    feature=next((f for f in _features(design) if _get(f,'id')==text),None)
    if feature:return feature_name(design,feature)
    if ':' in text:
        owner,suffix=text.split(':',1)
        placed=next((f for f in _features(design) if _get(f,'id')==owner),None)
        if placed and _get(placed,'kind')=='cavity':
            ids=list((_get(placed,'interface_nets',{}) or {}).keys())
            return interface_name(design,owner,suffix,index=ids.index(suffix) if suffix in ids else 0)
    return review_name(text)


def rule_name(rule):
    return _RULE_NAMES.get(rule,re.sub(r'_+',' ',str(rule or '')).capitalize())
